### ISSUE TITLE
A more machine-reasonable notion of unnecessary task avoidance

### ISSUE THREAD

**Issue by copumpkin** [sbt/sbt#1437]
There's currently `Tracking.inputChanged` and friends that you can throw inside a task definition that will avoid redoing things unless some condition is met.

I'm wondering whether that notion (a generally useful one) could be incorporated directly into the graph that sbt is handling behind the scenes, so it could know if it actually did anything or if the task's "redo-work" precondition failed. It'd also give us a handy way to override that logic in case the avoidance heuristic is imperfect.

Does this make sense as a feature?

**Comment by jsuereth** [comment]
Yes, we've debated it a bit. Leaving this open as enhancement but we'll need more information.

**Comment by eed3si9n** [comment]
Fixed in https://github.com/sbt/sbt/pull/7464
